kv/kvserver: TestStoreRangeMergeSlowUnabandonedFollower_WithSplit failed

kv/kvserver.TestStoreRangeMergeSlowUnabandonedFollower_WithSplit [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21042231?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21042231?buildTab=artifacts#/) on master @ [5320aa76299f5d3b52811406945caae7ee16e3a4](https://github.com/cockroachdb/cockroach/commits/5320aa76299f5d3b52811406945caae7ee16e3a4):


```
          | 	pkg/kv/kvserver/store_send.go:188
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes
          | 	pkg/kv/kvserver/stores.go:203
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
          | 	pkg/server/node.go:1669
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
          | 	pkg/server/node.go:1862
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func1
          | 	pkg/rpc/context.go:771
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12
          | 	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:60
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          | 	pkg/rpc/context.go:882
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
          | 	pkg/rpc/context.go:181
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          | 	pkg/rpc/context.go:882
          | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
          | 	pkg/rpc/auth.go:96
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          | 	pkg/rpc/context.go:882
          | github.com/cockroachdb/cockroach/pkg/rpc.gatewayRequestRecoveryInterceptor
          | 	pkg/rpc/metrics.go:553
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          | 	pkg/rpc/context.go:882
          | github.com/cockroachdb/cockroach/pkg/server.newGRPCServer.NewRequestMetricsInterceptor.func4
          | 	pkg/rpc/metrics.go:471
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          | 	pkg/rpc/context.go:882
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
          | 	pkg/rpc/context.go:138
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          | 	pkg/util/stop/stopper.go:351
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
          | 	pkg/rpc/context.go:136
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          | 	pkg/rpc/context.go:882
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func2
          | 	pkg/rpc/context.go:781
          | github.com/cockroachdb/cockroach/pkg/rpc.NewContext.ClientInterceptor.func9
          | 	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:185
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.getChainUnaryInvoker.func4
          | 	pkg/rpc/context.go:966
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func3
          | 	pkg/rpc/context.go:851
        Wraps: (5) refusing to transfer lease to (n3,s3):3 because target may need a Raft snapshot: replica in StateSnapshot
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *markers.withMark (4) *errbase.opaqueWrapper (5) *errutil.leafError
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/521e50704b845324c1d7d2447a7fd31d/logTestStoreRangeMergeSlowUnabandonedFollower_WithSplit1183901509
--- FAIL: TestStoreRangeMergeSlowUnabandonedFollower_WithSplit (13.70s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeSlowUnabandonedFollower_WithSplit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-59226

--- a/pkg/kv/kvserver/client_merge_test.go
+++ b/pkg/kv/kvserver/client_merge_test.go
@@ -2765,8 +2765,11 @@ func TestStoreRangeMergeSlowUnabandonedFollower_WithSplit(t *testing.T) {
 	// Transfer the lease on the new RHS to store2 and wait for it to apply. This
 	// will force its replica of the new RHS to become up to date, which
 	// indirectly tests that the replica GC queue cleans up both the LHS replica
-	// and the old RHS replica.
-	tc.TransferRangeLeaseOrFatal(t, *newRHSDesc, tc.Target(2))
+	// and the old RHS replica. The replica may still be in StateSnapshot, so we
+	// retry the lease transfer until it succeeds.
+	testutils.SucceedsSoon(t, func() error {
+		return tc.TransferRangeLease(*newRHSDesc, tc.Target(2))
+	})
 	testutils.SucceedsSoon(t, func() error {
 		rhsRepl, err := store2.GetReplica(newRHSDesc.RangeID)
 		if err != nil {

